### PR TITLE
Allow modifying admin of Ics20 contract

### PR DESCRIPTION
**Pull request by ethanfrey**
- [x] Use ADMIN rather then config
- [x] Support standard ADMIN query and update
- [x] Add migration support from current version to new version

### CODE REVIEW

File: contracts/cw20-ics20/src/contract.rs [hunk 1]
```diff
@@ -186,14 +192,54 @@ pub fn execute_allow(
...
+    // existing one is newer
+    if storage_version > version {
+        return Err(ContractError::CannotMigrateVersion {
+            previous_version: stored.version,
+        });
+    }
+
+    // Then, run the proper migration
+    // unsupported old version
+    if storage_version < "0.11.1".parse().map_err(from_semver)? {
```
Comment (maurolacy) [inline contracts/cw20-ics20/src/contract.rs:216]: Use a `MIGRATE_MIN_VERSION` constant?

File: contracts/cw20-ics20/src/contract.rs [hunk 2]
```diff
@@ -186,14 +192,54 @@ pub fn execute_allow(
...
+        return Err(ContractError::CannotMigrateVersion {
+            previous_version: stored.version,
+        });
+    }
+    // for 0.12.0-alpha1 or earlier, migrate from the config.gov_contract to ADMIN
+    if storage_version <= "0.12.0-alpha1".parse().map_err(from_semver)? {
```
Comment (maurolacy) [inline contracts/cw20-ics20/src/contract.rs:222]: Use a `MIGRATE_V1_VERSION` constant?

File: contracts/cw20-ics20/src/contract.rs [hunk 3]
```diff
@@ -186,14 +192,54 @@ pub fn execute_allow(
...
+
+    if storage_version < version {
+        // we don't need to save anything if migrating from the same version
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    }
+
+    Ok(Response::new())
+}
+
+fn from_semver(err: semver::Error) -> StdError {
```
Comment (maurolacy) [inline contracts/cw20-ics20/src/contract.rs:241]: Nice.

File: contracts/cw20-ics20/src/contract.rs [hunk 4]
```diff
@@ -251,9 +298,10 @@ pub fn query_channel(deps: Deps, id: String) -> StdResult<ChannelResponse> {
 
 fn query_config(deps: Deps) -> StdResult<ConfigResponse> {
     let cfg = CONFIG.load(deps.storage)?;
+    let admin = ADMIN.get(deps)?.unwrap_or_else(|| Addr::unchecked(""));
```
Comment (maurolacy) [inline contracts/cw20-ics20/src/contract.rs:301]: Nice. Wouldn't `"None"` be better here? I think we used that in the past.

File: contracts/cw20-ics20/src/contract.rs [hunk 5]
```diff
@@ -186,14 +192,54 @@ pub fn execute_allow(
...
+    // unsupported old version
+    if storage_version < "0.11.1".parse().map_err(from_semver)? {
+        return Err(ContractError::CannotMigrateVersion {
+            previous_version: stored.version,
+        });
+    }
+    // for 0.12.0-alpha1 or earlier, migrate from the config.gov_contract to ADMIN
+    if storage_version <= "0.12.0-alpha1".parse().map_err(from_semver)? {
+        // Do v1 migration
+        let old_config = crate::migrations::v1::CONFIG.load(deps.storage)?;
```
Comment (maurolacy) [inline contracts/cw20-ics20/src/contract.rs:224]: 👍🏼

File: contracts/cw20-ics20/src/contract.rs [hunk 6]
```diff
@@ -186,14 +192,54 @@ pub fn execute_allow(
...
+    }
+    // otherwise no migration (yet)
+
+    if storage_version < version {
+        // we don't need to save anything if migrating from the same version
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    }
+
+    Ok(Response::new())
+}
```
Comment (maurolacy) [inline contracts/cw20-ics20/src/contract.rs:239]: Not for now, but I wonder how much of this logic can be turned into a helper function.

**Review by maurolacy — APPROVED** [review]
LGTM.

File: contracts/cw20-ics20/src/contract.rs (hunk 4 above)
Comment (ethanfrey) [reply contracts/cw20-ics20/src/contract.rs:301]: In this contract, we don't allow None, so it doesn't map much to the ADMIN, I kind of force it

File: contracts/cw20-ics20/src/migrations.rs [hunk 7]
```diff
@@ -0,0 +1,16 @@
+// v1 format is anything older than 0.12.0
+pub mod v1 {
+    use schemars::JsonSchema;
+    use serde::{Deserialize, Serialize};
+
+    use cosmwasm_std::Addr;
+    use cw_storage_plus::Item;
+
+    #[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+    pub struct ConfigV1 {
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply contracts/cw20-ics20/src/migrations.rs:10]: I will do the other naming

File: contracts/cw20-ics20/src/contract.rs (hunk 6 above)
Comment (ethanfrey) [reply contracts/cw20-ics20/src/contract.rs:239]: some of this is in poe-contracts/tg-utils

but matching on versions would be a good thing to abstract out (later, when we have some good examples/experience)

**Pull request merged**